rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+		such that only Committers have to sign their commits.
+
+# Motivation
+[motivation]: #motivation
+
+Signing commits would allow cryptographically verifying that the source code
+hasn't been tampered by an untrusted party.
+
+This would ensure that when pulling from GitHub, or some other source,
+you can check if the source code has been signed by a trusted person.
+
+This is already done by Guix:
+- https://guix.gnu.org/en/blog/2020/securing-updates/
+- https://github.com/guix-mirror/guix/commits/master
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://guix.gnu.org/manual/en/html_node/Channel-Authentication.html
+
+It is also done by most conventional distros, however, they sign
+the output from the compilation rather than the source code, since
+they distribute the output from the compilation.
+
+This change would also allow using 3rd-party mirrors of Nixpkgs,
+since we can verify the source code independent of where it came from.
+
+This RFC was motivated by work during Summer of Nix, specifically,
+the Libre SoC maintainer requested that source integrity be
+maintained.
+
+# Detailed design
+[design]: #detailed-design
+
+## Signing the commits
+[signing]: #signing-the-commits
+
+The following will be required from 2022-03-01T00:00Z on,
+but is strongly recommended from now on.
+
+Currently merging is done through GitHub, which would not work in this case.
+Merging would have to happen outside of GitHub, so that you can sign the commits,
+since currently the API doesn't support signing commits with your own PGP
+key: https://github.com/cli/cli/issues/1318.
+
+We would make a script for merging pull requests instead, something like:
+```bash
+#!/bin/sh
+# nixpkgsmerge
+
+git switch master
+git pull
+gh pr checkout $1
+git switch master
+git merge -m "Merged #$1" --no-ff --gpg-sign -
+git push origin master
+gh pr close $1
+```
+Then
+```bash
+nixpkgsmerge 123
+```
+
+For mobile platforms, you would use [nix-on-droid](https://github.com/t184256/nix-on-droid)
+to run the script instead.
+
+A merge commit must always be made so that there is something to sign.
+
+A git hook like the one presented here is also recommended,
+so that unsigned commits aren't pushed by accident.
+This would be put in `/.git/hooks/pre-push` of the repository to
+be committed to.
+```bash
+#!/bin/sh
+
+# Inspiration taken from https://github.com/guix-mirror/guix/blob/master/etc/git/pre-push.
+# Put this in .git/hooks/pre-push.
+set -xe
+
+# This is the "empty hash" used by Git when pushing a branch deletion.
+z40=0000000000000000000000000000000000000000
+
+while read local_ref local_hash remote_ref remote_hash
+do
+  # Branch deletion; no signature to verify.
+  if [ "$local_hash" != $z40 ]
+  then
+    # Check repositories under github.com/NixOS.
+    case "$2" in
+        *github.com*NixOS/*)
+      # TODO: More thorough check.
+      git verify-commit HEAD
```

Comment:
How do we handle expired keys? What does that mean for us? Do we accept that as "accident" that can happen and if so when would we reject commits that have expired keys on them but recently pushed to master?

Reply:
Expired keys can't be used for verifying trust at all. As I explained in another GitHub conversation, such accidents would mean the history needs to be rewritten, but since nobody would update to the untrusted commit, there is very little risk of breakage for downstream users.